### PR TITLE
Compute type of inherited constructor correctly

### LINKED ISSUE #1246: __init__ of subclass of a Generic class uses the generic type instead of the specified one
Code to repro:

```
#!/usr/bin/env python
from typing import Generic, TypeVar

T = TypeVar("T")

class Test(Generic[T]):

    def __init__(self, a):
        # type: (T) -> None
        self.a = a

class StrTest(Test[str]):
    pass

s = StrTest("str1")
```

I would expect there to be no error here, but mypy outputs the following:

```
test.py:15: error: Argument 1 to "StrTest" has incompatible type "str"; expected "T"
```

### PR DESCRIPTION
**Pull request by rwbarton**
Fixes #1246.

### CODE REVIEW

**Comment by JukkaL** [comment]
Looks good -- assuming that the build passes :-)

File: mypy/checkmember.py [hunk 1]
```diff
@@ -359,6 +359,8 @@ def type_object_type(info: TypeInfo, builtin_type: Callable[[str], Instance]) ->
 def type_object_type_from_function(init_or_new: FuncBase, info: TypeInfo,
                                    fallback: Instance) -> FunctionLike:
     signature = method_type_with_fallback(init_or_new, fallback)
+    signature = cast(FunctionLike,
+                     map_type_from_supertype(signature, info, init_or_new.info))
```
Comment (rwbarton) [inline mypy/checkmember.py]: This really deserves a comment, doesn't it.

Comment (JukkaL) [inline mypy/checkmember.py]: Yeah, good point. Maybe some reasoning about why the cast is safe and a general description.

**Comment by rwbarton** [comment]
Added some comments explaining what is going on here. I don't think the cast really needs a special mention once there is a comment giving the general picture.

**Comment by JukkaL** [comment]
Rebased and pushed.

**Pull request closed without merging**